File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/DP/StairUniqueWays.ts
function stairUniqueWaysBU(n: number): number {
	if (n <= 2) return n;
	let first = 1,
		second = 2,
		temp = 0;

	for (let i = 3; i <= n; i++) {
		temp = first + second;
		first = second;
		second = temp;
	}
	return temp;
}

function stairUniqueWaysBU2(n: number): number {
	if (n < 2) return n;

	const ways: number[] = new Array(n);
	ways[0] = 1;
	ways[1] = 2;

	for (let i = 2; i < n; i++) {
		ways[i] = ways[i - 1] + ways[i - 2];
	}

	return ways[n - 1];
}

console.log("Unique ways to reach the top: " + stairUniqueWaysBU(4));
console.log("Unique ways to reach the top: " + stairUniqueWaysBU2(4));

/*
Unique ways to reach the top: 5
Unique ways to reach the top: 5
*/

File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/DP/StairUniqueWays.swift
import Foundation

func stairUniqueWaysBU(_ n : Int) -> Int {
	if (n <= 2) {
		return n
	}

	var first : Int = 1
	var second : Int = 2
	var temp : Int = 0
	var i : Int = 3
	while (i <= n) {
		temp = first + second
		first = second
		second = temp
		i += 1
	}
	return temp
}

func stairUniqueWaysBU2(_ n : Int) -> Int {
	if (n < 2) {
		return n
	}
	var ways : [Int] = Array(repeating: 0, count: n)
	ways[0] = 1
	ways[1] = 2

	var i : Int = 2
	while (i < n) {
		ways[i] = ways[i - 1] + ways[i - 2]
		i += 1
	}
	return ways[n - 1]
}

// Testing code.
print("Unique way to reach top::" , stairUniqueWaysBU(4))
print("Unique way to reach top::", stairUniqueWaysBU2(4))

/*
Unique way to reach top:: 5
Unique way to reach top:: 5
*/